canary/consumer-java/src/main/java/com/amazon/kinesis/video/canary/consumer/CanaryListFragmentWorker.java

```diff
@@ -0,0 +1,107 @@
+                                                  final Regions region,
+                                                  final AmazonKinesisVideo amazonKinesisVideo,
+                                                  final FragmentSelector fragmentSelector) {
+        final GetDataEndpointRequest request = new GetDataEndpointRequest()
+                .withAPIName(APIName.LIST_FRAGMENTS).withStreamName(streamName);
+        final String endpoint = amazonKinesisVideo.getDataEndpoint(request).getDataEndpoint();
+        
+        return new CanaryListFragmentWorker(
+                streamName, credentialsProvider, endpoint, region, fragmentSelector);
+    }
+
+    @Override
+    public List<CanaryFragment> call() {
+        List<CanaryFragment> fragments = new ArrayList<>();
+        try {
+            log.info("Start CanaryListFragment worker on stream {}", streamName);
+            
+            ListFragmentsRequest request = new ListFragmentsRequest()
+                    .withStreamName(streamName).withFragmentSelector(fragmentSelector).withMaxResults(fragmentsPerRequest);
+
+            ListFragmentsResult result = amazonKinesisVideoArchivedMedia.listFragments(request);
+            
+
+            log.info("List Fragments called on stream {} response {} request ID {}",
+                    streamName,
+                    result.getSdkHttpMetadata().getHttpStatusCode(),
+                    result.getSdkResponseMetadata().getRequestId());
+
+            for (Fragment f : result.getFragments()) {
+                fragments.add(new CanaryFragment(f));
+            }
+            String nextToken = result.getNextToken();
+
+            /* If result is truncated, keep making requests until nextToken is empty */
+            while (nextToken != null) {
+                request = new ListFragmentsRequest()
+                        .withStreamName(streamName).withNextToken(nextToken);
+                result = amazonKinesisVideoArchivedMedia.listFragments(request);
+
+                for (Fragment f : result.getFragments()) {
+                    fragments.add(new CanaryFragment(f));
+                }
+                nextToken = result.getNextToken();
+            }
+
+            // Fragments in the return of a listFragments call are in no particular order,
+            // so let's sort them by ascending fragment number
+            fragments.sort(Comparator.comparing(CanaryFragment::getFragmentNumberInt));
+
+            for (CanaryFragment cf : fragments) {
+                log.info("Retrieved fragment number {} ", cf.getFragment().getFragmentNumber());
+            }
+        } catch (Exception e) {
+            log.error("Failure in CanaryListFragmentWorker for streamName {} {}", streamName, e.toString());
+            throw e;
+        } finally {
+                        log.info("Retrieved {} Fragments and exiting CanaryListFragmentWorker for stream {}", fragments.size(), streamName);
+            return fragments;
```

Comment:
This function has return type `List<CanaryFragment>`. What would you prefer to return if not an empty list?

Reply:
Also, there is a chance that the list is partially populated before an exception is thrown, this would allow for that partial list to be returned.

If we want to diverge so much from the parser library example this whole class is a copy of, would you like me to fix up that example too?

Reply:
It can return empty list. Returning from `finally` block is not intuitive. `finally` is a block that is used to run cleanup or an operation that should be executed irrespective of exception. It is generally advisable to have return at the end of the function and not within these blocks. Parser library is a standalone sample and  although we are taking pieces of code from those samples to set this up, we should ensure this code follows good practices. We can look into cleaning those up wherever relevant in the future.